source/time_tables/thermometer.cpp

```diff
@@ -130,14 +132,72 @@ void Thermometer::generate_thermometer(const type::Route* route) {
     generate_thermometer(std::vector<vector_idx>(stop_point_lists.begin(), stop_point_lists.end()));
 }
 
+// Define types for next function 'generate_topological_thermometer'
+struct VertexProperties
+{
+    type::idx_t idx;
+    VertexProperties() : idx(0) {}
+    VertexProperties(type::idx_t const& i): idx(i) {}
+};
+typedef boost::adjacency_list< boost::vecS, boost::vecS, boost::directedS, boost::property<boost::vertex_bundle_t, VertexProperties> > Graph;
+typedef boost::graph_traits<Graph>::vertex_descriptor Vertex;
+
+// Generate an topologicaly exact thermometer
+// return false if topological sort fail (because of a cycle)
+// return true if succeed
+bool Thermometer::generate_topological_thermometer(const std::vector<vector_idx> &stop_point_lists) {
+
+    Graph stop_point_graph;
+    std::map<type::idx_t, Vertex> vertex_map;
+    for (const auto& v: stop_point_list) {
+        bool first = true;
+        Vertex previous_vx;
+        for (auto sp : v) {
+            Vertex current_vx;
+
+            // We don't add a vertex twice
+            auto it = vertex_map.find(sp);
+            if (it == vertex_map.end()) {
+                current_vx = boost::add_vertex(VertexProperties(sp), stop_point_graph);
+                vertex_map[sp] = current_vx;
+            } else {
+	        current_vx = it->second;
+	    }
+	    if (!first) {
```

Comment:
Do you prefer a map of pointers ?

Else, there is a solution with labeled_graph
But that sound tricky and slow in comparaison to my O(1) map index :)